pkg/apis/sparkoperator.k8s.io/v1beta2/types.go

```diff
@@ -0,0 +1,569 @@
+	TerminationTime metav1.Time `json:"terminationTime,omitempty"`
+	// DriverInfo has information about the driver.
+	DriverInfo DriverInfo `json:"driverInfo"`
+	// AppState tells the overall application state.
+	AppState ApplicationState `json:"applicationState,omitempty"`
+	// ExecutorState records the state of executors by executor Pod names.
+	ExecutorState map[string]ExecutorState `json:"executorState,omitempty"`
+	// ExecutionAttempts is the total number of attempts to run a submitted application to completion.
+	// Incremented upon each attempted run of the application and reset upon invalidation.
+	ExecutionAttempts int32 `json:"executionAttempts,omitempty"`
+	// SubmissionAttempts is the total number of attempts to submit an application to run.
+	// Incremented upon each attempted submission of the application and reset upon invalidation and rerun.
+	SubmissionAttempts int32 `json:"submissionAttempts,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SparkApplicationList carries a list of SparkApplication objects.
+type SparkApplicationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []SparkApplication `json:"items,omitempty"`
+}
+
+// Dependencies specifies all possible types of dependencies of a Spark application.
+type Dependencies struct {
+	// Jars is a list of JAR files the Spark application depends on.
+	// Optional.
+	Jars []string `json:"jars,omitempty"`
+	// Files is a list of files the Spark application depends on.
+	// Optional.
+	Files []string `json:"files,omitempty"`
+	// PyFiles is a list of Python files the Spark application depends on.
+	// Optional.
+	PyFiles []string `json:"pyFiles,omitempty"`
+	// JarsDownloadDir is the location to download jars to in the driver and executors.
+	JarsDownloadDir *string `json:"jarsDownloadDir,omitempty"`
+	// FilesDownloadDir is the location to download files to in the driver and executors.
+	FilesDownloadDir *string `json:"filesDownloadDir,omitempty"`
+	// DownloadTimeout specifies the timeout in seconds before aborting the attempt to download
+	// and unpack dependencies from remote locations into the driver and executor pods.
+	// +kubebuilder:validation:Minimum=1
+	DownloadTimeout *int32 `json:"downloadTimeout,omitempty"`
+	// MaxSimultaneousDownloads specifies the maximum number of remote dependencies to download
+	// simultaneously in a driver or executor pod.
+	// +kubebuilder:validation:Minimum=1
+	MaxSimultaneousDownloads *int32 `json:"maxSimultaneousDownloads,omitempty"`
+}
+
+// SparkPodSpec defines common things that can be customized for a Spark driver or executor pod.
+// TODO: investigate if we should use v1.PodSpec and limit what can be set instead.
+type SparkPodSpec struct {
+	// Cores is the number of CPU cores to request for the pod.
+	// Optional.
+	// +kubebuilder:validation:Minimum=0
```

Comment:
Can we do `+kubebuilder:validation:Minimum=1`?

Reply:
Ah, right, `ExclusiveMinimum` isn't necessary now that this is an integer. 👍